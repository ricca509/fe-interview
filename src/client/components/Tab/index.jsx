import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const Tab = ({ title, tabIndex, onClick, isSelected = false }) => {
  const onTabClick = event => {
    event.preventDefault();
    onClick(tabIndex);
  };

  const classes = classNames({
    tab: true,
    'tab-selected': isSelected
  });

  return (
    <li className={classes}>
      <a href="#" onClick={onTabClick}>
        {title}
      </a>
    </li>
  );
};

Tab.propTypes = {
  title: PropTypes.string.isRequired,
  tabIndex: PropTypes.number.isRequired,
  onClick: PropTypes.func.isRequired,
  isSelected: PropTypes.bool
};

export default Tab;