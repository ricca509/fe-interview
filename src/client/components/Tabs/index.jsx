import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Tab from '../Tab';

class Tabs extends Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedTabIdx: this.props.preSelectedTabIndex || 0
    };

    this.onTabSelect = this.onTabSelect.bind(this);
  }

  onTabSelect(selectedTabIdx) {
    this.setState(() => ({
      selectedTabIdx
    }));
  }

  renderTabHeaders() {
    const { children } = this.props;
    const { selectedTabIdx } = this.state;

    return React.Children
      .toArray(children)
      .map((tab, idx) => {
        return React.cloneElement(tab, {
          tabIndex: idx,
          isSelected: selectedTabIdx === idx,
          onClick: this.onTabSelect
        });
      });
  }

  renderTabContent() {
    const { children } = this.props;
    const { selectedTabIdx } = this.state;

    return React.Children
      .toArray(children)
      .filter((tab, idx) => selectedTabIdx === idx)
      .map(tab => tab.props.children)
  }

  render() {
    const { children } = this.props;
    return (
      <div className="tabs">
        <ul>
          {this.renderTabHeaders()}
        </ul>
        <div className="tab-content">
          {this.renderTabContent()}
        </div>
      </div>
    );
  }
}

Tabs.propTypes = {
  children: PropTypes.node.isRequired,
  preSelectedTabIndex: PropTypes.number
};

export default Tabs;