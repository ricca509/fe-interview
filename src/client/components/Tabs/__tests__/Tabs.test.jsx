import React from 'react';
import Tabs from '../';
import Tab from '../../Tab';
import { shallow } from 'enzyme';

describe('Tabs component', ()=> {
  it('should render a list of `Tab` with the right props and one selected content', () => {
    const wrapper = shallow(
      <Tabs>
        <Tab title="Tab1">
          <div>content1</div>
        </Tab>
        <Tab title="Tab2">
          <div>content2</div>
        </Tab>
      </Tabs>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render the right selected content when a `Tab` is clicked', () => {
    const newlySelectedTabIndex = 1;
    const wrapper = shallow(
      <Tabs>
        <Tab title="Tab1">
          <div>content1</div>
        </Tab>
        <Tab title="Tab2">
          <div>content2</div>
        </Tab>
      </Tabs>
    );

    wrapper.find('Tab').at(1).simulate('click', newlySelectedTabIndex);

    expect(wrapper).toMatchSnapshot();
  });
});