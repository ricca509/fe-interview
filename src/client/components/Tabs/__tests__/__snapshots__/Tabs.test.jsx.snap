// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component should render a list of \`Tab\` with the right props and one selected content 1`] = `
<div
  className="tabs"
>
  <ul>
    <Tab
      isSelected={true}
      key=".0"
      onClick={[Function]}
      tabIndex={0}
      title="Tab1"
    >
      <div>
        content1
      </div>
    </Tab>
    <Tab
      isSelected={false}
      key=".1"
      onClick={[Function]}
      tabIndex={1}
      title="Tab2"
    >
      <div>
        content2
      </div>
    </Tab>
  </ul>
  <div
    className="tab-content"
  >
    <div>
      content1
    </div>
  </div>
</div>
`;

exports[`Tabs component should render the right selected content when a \`Tab\` is clicked 1`] = `
<div
  className="tabs"
>
  <ul>
    <Tab
      isSelected={false}
      key=".0"
      onClick={[Function]}
      tabIndex={0}
      title="Tab1"
    >
      <div>
        content1
      </div>
    </Tab>
    <Tab
      isSelected={true}
      key=".1"
      onClick={[Function]}
      tabIndex={1}
      title="Tab2"
    >
      <div>
        content2
      </div>
    </Tab>
  </ul>
  <div
    className="tab-content"
  >
    <div>
      content2
    </div>
  </div>
</div>
`;
